system/lib/wasmfs/wasmfs.cpp

```diff
@@ -62,9 +62,46 @@ extern "C" void wasmfs_flush(void) {
   // Flush musl libc streams.
   fflush(0);
 
-  // Flush our own streams. TODO: flush all backends.
+  // Flush our own streams.
   (void)SpecialFiles::getStdout()->locked().flush();
   (void)SpecialFiles::getStderr()->locked().flush();
+
+  std::vector<std::shared_ptr<Directory>> toFlush;
+  toFlush.push_back(wasmFS.getRootDirectory());
+
+  while (!toFlush.empty()) {
+    auto dir = toFlush.back();
+    toFlush.pop_back();
+
+    auto lockedDir = dir->locked();
+    Directory::MaybeEntries entries = lockedDir.getEntries();
+    if (int err = entries.getError()) {
+#ifndef NDEBUG
+      std::string errorMessage =
+        "Non-fatal error code " + std::to_string(err) +
+        " while flushing directory: " + lockedDir.getName(dir);
+      emscripten_console_error(errorMessage.c_str());
+#endif
+      continue;
+    }
+    for (auto entry : *entries) {
```

Comment:
I've isolated the error to this line. The test fails when I use `for (auto entry : *entries) {` but passes when I use `for (auto entry : *lockedDir.getEntries()) {`. I'm reasonably sure that these two lines should be the same thing since I assign `Directory::MaybeEntries entries = lockedDir.getEntries();` on Line 77. Is there some specific reason the first version fails?

Reply:
Not sure why this is failing, but one option would be to edit system_libs.py to pass `-fsanitize=address` when building WasmFS. AddressSanitizer will help catch any memory errors the code might have.

Out of curiosity, does it work when you do `for (auto& entry : ...)`? (note the `auto&` making `entry` a reference rather than a copied value)

Reply:
I try the `auto&` line and it still worked for `*lockedDir.getEntries()` but not `*entries`.

When I tried using AddressSanitizer, there weren't any errors raised during the compilation (I also tried running the exact command from `--verbose`), but the test ended up having an error inside of it that wasn't shown even in `--verbose` mode. The test fails because it logs to `/report_result?exit:1` `stdout` without any additional logs. This is unusual because usually the test becomes unresponsive, ending with something like `no http response`. Is there another way besides `--verbose` to check on what's happening inside the test?

Reply:
Oh right, I forgot that you don't need to edit system_libs.py to use ASan. Too bad that didn't turn up anything immediately useful.

Maybe the different behavior you're seeing is from `-sEXIT_RUNTIME`? IIRC, using ASan enables EXIT_RUNTIME implicitly, so maybe if you set it explicitly, you'll be able to see the same thing? Just a guess.